### PR TITLE
planner: fix index out of range in constructIndexJoinInnerSideTask

### LINKED ISSUE #54535: SQL failed with "ERROR 1105 (HY000): runtime error: index out of range [1] with length 1" when tidb_enable_inl_join_inner_multi_pattern is ON
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```SQL
set GLOBAL tidb_enable_inl_join_inner_multi_pattern='ON';
create table ta(a1 int, a2 int, a3 int, index idx_a(a1));
create table tb(b1 int, b2 int, b3 int, index idx_b(b1));
explain SELECT /*+ inl_join(tmp) */ * FROM ta, (SELECT b1, COUNT(b3) AS cnt FROM tb GROUP BY b1, b2) as tmp where ta.a1 = tmp.b1;
```

### 2. What did you expect to see? (Required)

Output physical plan like

```
+-----------------------------------+---------+-----------+---------------------------+-----------------------------------------------------------------------------------------------------------------+
| id                                | estRows | task      | access object             | operator info                                                                                                   |
+-----------------------------------+---------+-----------+---------------------------+-----------------------------------------------------------------------------------------------------------------+
| Projection_9                      | 3.00    | root      |                           | test.ta.a1, test.ta.a2, test.ta.a3, test.tb.b1, Column#9                                                        |
| └─IndexJoin_16                    | 3.00    | root      |                           | inner join, inner:HashAgg_14, outer key:test.ta.a1, inner key:test.tb.b1, equal cond:eq(test.ta.a1, test.tb.b1) |
|   ├─TableReader_43(Build)         | 3.00    | root      |                           | data:Selection_42                                                                                               |
|   │ └─Selection_42                | 3.00    | cop[tikv] |                           | not(isnull(test.ta.a1))                                                                                         |
|   │   └─TableFullScan_41          | 3.00    | cop[tikv] | table:ta                  | keep order:false, stats:pseudo                                                                                  |
|   └─HashAgg_14(Probe)             | 19.16   | root      |                           | group by:test.tb.b1, test.tb.b2, funcs:count(Column#11)->Column#9, funcs:firstrow(test.tb.b1)->test.tb.b1       |
|     └─IndexLookUp_15              | 19.16   | root      |                           |                                                                                                                 |
|       ├─Selection_12(Build)       | 3.00    | cop[tikv] |                           | not(isnull(test.tb.b1))                                                                                         |
|       │ └─IndexRangeScan_10       | 3.00    | cop[tikv] | table:tb, index:idx_b(b1) | range: decided by [eq(test.tb.b1, test.ta.a1)], keep order:false, stats:pseudo                                  |
|       └─HashAgg_13(Probe)         | 19.16   | cop[tikv] |                           | group by:test.tb.b1, test.tb.b2, funcs:count(test.tb.b3)->Column#11                                             |
|         └─TableRowIDScan_11       | 3.00    | cop[tikv] | table:tb                  | keep order:false, stats:pseudo                                                                                  |
+-----------------------------------+---------+-----------+---------------------------+-----------------------------------------------------------------------------------------------------------------+
```

### 3. What did you see instead (Required)

```
ERROR 1105 (HY000): runtime error: index out of range [1] with length 1
```

### 4. What is your TiDB version? (Required)

v8.2.0-alpha-557-g07bc0d09e8

### PR DESCRIPTION
**Pull request by joechenrh**
### What problem does this PR solve?

Issue Number: close #54535

Problem Summary:

When set system variable `tidb_enable_inl_join_inner_multi_pattern` to `ON`, the following SQL will get an error.

```SQL
set GLOBAL tidb_enable_inl_join_inner_multi_pattern='ON';
create table ta(a1 int, a2 int, a3 int, index idx_a(a1));
create table tb(b1 int, b2 int, b3 int, index idx_b(b1));
explain SELECT /*+ inl_join(tmp) */ * FROM ta, (SELECT b1, COUNT(b3) AS cnt FROM tb GROUP BY b1, b2) as tmp where ta.a1 = tmp.b1;

ERROR 1105 (HY000): runtime error: index out of range [1] with length 1
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -1455,12 +1455,13 @@ func (p *LogicalJoin) constructIndexJoinInnerSideTask(dsCopTask *CopTask, ds *Da
 		ds.TableInfo.GetPartitionInfo() == nil {
 		if len(path.IdxCols) < len(groupByCols) {
 			preferStream = false
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply pkg/planner/core/exhaust_physical_plans.go:1457]: You may refer to the implementation in pkg/planner/core/issuetest/planner_issue_test.go.

File: pkg/planner/core/issuetest/planner_issue_test.go [hunk 2]
```diff
@@ -86,3 +86,33 @@ func Test53726(t *testing.T) {
...
+	// The optimizer should choose IndexJoin
+	var ok bool
+	for {
+		_, ok = p.(*core.PhysicalIndexJoin)
+		if ok || len(p.(base.PhysicalPlan).Children()) == 0 {
+			break
+		}
+		p = p.(base.PhysicalPlan).Children()[0]
+	}
+	require.True(t, ok)
```
Comment (hawkingrei) [inline pkg/planner/core/issuetest/planner_issue_test.go:117]: You can use ```tk.MustQuery("explain xxx")``` to check whether the planner is right.

File: pkg/planner/core/issuetest/planner_issue_test.go [hunk 3]
```diff
@@ -86,3 +86,33 @@ func Test53726(t *testing.T) {
 			"  └─TableReader_11 2.00 root  data:TableFullScan_10",
 			"    └─TableFullScan_10 2.00 cop[tikv] table:t7 keep order:false"))
 }
+
+func Test54535(t *testing.T) {
```
Comment (winoros) [inline pkg/planner/core/issuetest/planner_issue_test.go:90]:
```suggestion
func TestIssue54535(t *testing.T) {
```

**Review by winoros — COMMENTED** [review]
Just align with other test case name

**Comment by hawkingrei** [comment]
@joechenrh Please sync the code with master.

**Review by lance6716 — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Pull request merged**